File: telegram-bot-game/sonnet-4-5/bot.js
require('dotenv').config();
const TelegramBot = require('node-telegram-bot-api');

// Validate environment variables
if (!process.env.TELEGRAM_BOT_TOKEN) {
  console.error('Error: TELEGRAM_BOT_TOKEN is not set in .env file');
  process.exit(1);
}

// Initialize bot
const bot = new TelegramBot(process.env.TELEGRAM_BOT_TOKEN, { polling: true });

// Game state storage (in-memory)
// Key: chatId, Value: { min, max, target, attempts }
const games = new Map();

// Default range
const DEFAULT_MIN = 1;
const DEFAULT_MAX = 100;

/**
 * Initialize a new game for a chat
 */
function initializeGame(chatId, min = DEFAULT_MIN, max = DEFAULT_MAX) {
  const target = Math.floor(Math.random() * (max - min + 1)) + min;
  games.set(chatId, {
    min,
    max,
    target,
    attempts: 0
  });
  return { min, max };
}

/**
 * Get game state for a chat
 */
function getGame(chatId) {
  return games.get(chatId);
}

/**
 * Check if game exists for a chat
 */
function hasGame(chatId) {
  return games.has(chatId);
}

/**
 * Delete game for a chat
 */
function deleteGame(chatId) {
  games.delete(chatId);
}

// Command: /start
bot.onText(/\/start/, (msg) => {
  const chatId = msg.chat.id;
  const { min, max } = initializeGame(chatId);

  bot.sendMessage(
    chatId,
    `Welcome to the Number Guessing Game!\n\n` +
    `I've picked a number between ${min} and ${max}.\n` +
    `Try to guess it! Just send me a number.\n\n` +
    `Commands:\n` +
    `/newgame - Start a new game\n` +
    `/range [min] [max] - Set a custom range (e.g., /range 1 1000)\n` +
    `/help - Show this help message`
  );
});

// Command: /newgame
bot.onText(/\/newgame/, (msg) => {
  const chatId = msg.chat.id;
  const game = getGame(chatId);

  // Use existing range if available, otherwise use defaults
  const min = game ? game.min : DEFAULT_MIN;
  const max = game ? game.max : DEFAULT_MAX;

  initializeGame(chatId, min, max);

  bot.sendMessage(
    chatId,
    `New game started!\n` +
    `I've picked a new number between ${min} and ${max}.\n` +
    `Good luck!`
  );
});

// Command: /range
bot.onText(/\/range(?:\s+(\d+)\s+(\d+))?/, (msg, match) => {
  const chatId = msg.chat.id;

  // Check if parameters were provided
  if (!match[1] || !match[2]) {
    bot.sendMessage(
      chatId,
      `Usage: /range [min] [max]\n\n` +
      `Example: /range 1 1000\n\n` +
      `This will start a new game with numbers between 1 and 1000.`
    );
    return;
  }

  const min = parseInt(match[1]);
  const max = parseInt(match[2]);

  // Validate range
  if (min >= max) {
    bot.sendMessage(
      chatId,
      `Invalid range! The minimum value must be less than the maximum value.\n` +
      `Please try again with a valid range.`
    );
    return;
  }

  if (max - min < 1) {
    bot.sendMessage(
      chatId,
      `Invalid range! The range must have at least 2 numbers.`
    );
    return;
  }

  // Initialize game with new range
  initializeGame(chatId, min, max);

  bot.sendMessage(
    chatId,
    `Great! I've picked a number between ${min} and ${max}.\n` +
    `Start guessing!`
  );
});

// Command: /help
bot.onText(/\/help/, (msg) => {
  const chatId = msg.chat.id;

  bot.sendMessage(
    chatId,
    `*Number Guessing Game - Help*\n\n` +
    `*How to play:*\n` +
    `I'll pick a random number in a range, and you try to guess it.\n` +
    `I'll tell you if your guess is too high or too low.\n\n` +
    `*Commands:*\n` +
    `/start - Start a new game (default range: 1-100)\n` +
    `/newgame - Start a new game with the same range\n` +
    `/range [min] [max] - Set a custom range and start a new game\n` +
    `  Example: /range 1 1000\n` +
    `/help - Show this help message\n\n` +
    `*Tips:*\n` +
    `- Just send a number to make a guess\n` +
    `- Try to guess in as few attempts as possible!\n` +
    `- Use binary search strategy for faster wins`,
    { parse_mode: 'Markdown' }
  );
});

// Handle regular messages (guesses)
bot.on('message', (msg) => {
  // Skip if it's a command
  if (msg.text && msg.text.startsWith('/')) {
    return;
  }

  const chatId = msg.chat.id;
  const text = msg.text;

  // Check if game exists
  if (!hasGame(chatId)) {
    bot.sendMessage(
      chatId,
      `No active game! Start a new game with /start or /newgame`
    );
    return;
  }

  // Validate input is a number
  const guess = parseInt(text);

  if (isNaN(guess)) {
    bot.sendMessage(
      chatId,
      `Please send a valid number! Your guess: "${text}" is not a number.`
    );
    return;
  }

  const game = getGame(chatId);

  // Check if guess is within range
  if (guess < game.min || guess > game.max) {
    bot.sendMessage(
      chatId,
      `Please guess a number between ${game.min} and ${game.max}!`
    );
    return;
  }

  // Increment attempts
  game.attempts++;

  // Check if correct
  if (guess === game.target) {
    const attempts = game.attempts;
    const attemptsText = attempts === 1 ? '1 attempt' : `${attempts} attempts`;

    bot.sendMessage(
      chatId,
      `Congratulations! You guessed it!\n\n` +
      `The number was ${game.target}.\n` +
      `You took ${attemptsText} to win.\n\n` +
      `Would you like to play again? Use /newgame to start a new game!`
    );

    // Keep the game state so /newgame can use the same range
    // but reset it with a new target
    const { min, max } = game;
    initializeGame(chatId, min, max);
  }
  // Check if too low
  else if (guess < game.target) {
    bot.sendMessage(
      chatId,
      `Too low! Try a higher number.\n` +
      `Attempts: ${game.attempts}`
    );
  }
  // Must be too high
  else {
    bot.sendMessage(
      chatId,
      `Too high! Try a lower number.\n` +
      `Attempts: ${game.attempts}`
    );
  }
});

// Error handling
bot.on('polling_error', (error) => {
  console.error('Polling error:', error.code, error.message);
});

bot.on('error', (error) => {
  console.error('Bot error:', error);
});

console.log('Bot is running...');
console.log('Press Ctrl+C to stop.');

File: telegram-bot-game/gpt-5-2-codex-high/bot.py
import logging
import os
import random
from dataclasses import dataclass

from telegram import Update
from telegram.ext import ApplicationBuilder, CommandHandler, ContextTypes, MessageHandler, filters

DEFAULT_MIN = 1
DEFAULT_MAX = 100


@dataclass
class GameState:
    min_value: int
    max_value: int
    target: int
    attempts: int
    active: bool


GAMES: dict[tuple[int, int], GameState] = {}


def _game_key(update: Update) -> tuple[int, int]:
    chat_id = update.effective_chat.id if update.effective_chat else 0
    user_id = update.effective_user.id if update.effective_user else 0
    return chat_id, user_id


def _create_game(min_value: int, max_value: int) -> GameState:
    return GameState(
        min_value=min_value,
        max_value=max_value,
        target=random.randint(min_value, max_value),
        attempts=0,
        active=True,
    )


def _ensure_game(key: tuple[int, int]) -> GameState:
    if key not in GAMES:
        GAMES[key] = _create_game(DEFAULT_MIN, DEFAULT_MAX)
    return GAMES[key]


async def start(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    key = _game_key(update)
    GAMES[key] = _create_game(DEFAULT_MIN, DEFAULT_MAX)
    await update.message.reply_text(
        "Welcome to Number Guess!\n"
        "I'm thinking of a number between 1 and 100. Send a guess and I'll say higher or lower.\n"
        "Commands: /newgame, /range <min> <max>, /help"
    )


async def help_command(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    await update.message.reply_text(
        "Number Guess commands:\n"
        "/start - start a new game in the default range (1-100)\n"
        "/newgame - start a new game with the current range\n"
        "/range <min> <max> - set a new range and start a game\n"
        "/help - show this help\n"
        "Send any number to make a guess."
    )


async def newgame(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    key = _game_key(update)
    if key in GAMES:
        current = GAMES[key]
        GAMES[key] = _create_game(current.min_value, current.max_value)
    else:
        GAMES[key] = _create_game(DEFAULT_MIN, DEFAULT_MAX)
    game = GAMES[key]
    await update.message.reply_text(
        f"New game started. Guess a number between {game.min_value} and {game.max_value}."
    )


async def set_range(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if len(context.args) != 2:
        await update.message.reply_text("Usage: /range <min> <max>")
        return
    try:
        min_value = int(context.args[0])
        max_value = int(context.args[1])
    except ValueError:
        await update.message.reply_text("Range values must be whole numbers.")
        return
    if min_value >= max_value:
        await update.message.reply_text("Range must have min < max.")
        return
    key = _game_key(update)
    GAMES[key] = _create_game(min_value, max_value)
    await update.message.reply_text(
        f"Range set to {min_value}-{max_value}. New game started!"
    )


async def handle_guess(update: Update, context: ContextTypes.DEFAULT_TYPE) -> None:
    if not update.message or not update.message.text:
        return
    key = _game_key(update)
    game = _ensure_game(key)
    if not game.active:
        await update.message.reply_text(
            "That game is over. Start a new one with /newgame or set a range with /range."
        )
        return
    text = update.message.text.strip()
    try:
        guess = int(text)
    except ValueError:
        await update.message.reply_text("Please send a whole number.")
        return
    if guess < game.min_value or guess > game.max_value:
        await update.message.reply_text(
            f"Your guess is out of range. Try {game.min_value}-{game.max_value}."
        )
        return
    game.attempts += 1
    if guess < game.target:
        await update.message.reply_text("Higher.")
    elif guess > game.target:
        await update.message.reply_text("Lower.")
    else:
        game.active = False
        await update.message.reply_text(
            f"Correct! You got it in {game.attempts} attempts.\n"
            "Want to play again? Use /newgame or set a new /range."
        )


def main() -> None:
    logging.basicConfig(
        format="%(asctime)s %(levelname)s %(name)s %(message)s",
        level=logging.INFO,
    )
    token = os.environ.get("TELEGRAM_BOT_TOKEN")
    if not token:
        raise RuntimeError("TELEGRAM_BOT_TOKEN is not set.")
    app = ApplicationBuilder().token(token).build()
    app.add_handler(CommandHandler("start", start))
    app.add_handler(CommandHandler("help", help_command))
    app.add_handler(CommandHandler("newgame", newgame))
    app.add_handler(CommandHandler("range", set_range))
    app.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, handle_guess))
    app.run_polling()


if __name__ == "__main__":
    main()
